### DISCUSSION TITLE
Why is fetching from the local API within getServerSideProps not recommended?

### DISCUSSION THREAD

**Discussion by kevanstannard** [vercel/next.js#23789] [Help]
In the documentation for `getServerSideProps` it's recommended to not to call the local API using fetch.
 
https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

![image](https://user-images.githubusercontent.com/937077/113950000-a69dd080-9853-11eb-9ae9-9ae198751d50.png)

I'm wondering why this is the recommended approach?

Is there a NextJS technical concern with this approach (other than being a little slower)?

For context, I am considering a future scenario where the API may be a separate server from NextJS application, so using fetches to API URLs for all requests now (both server side and client side) would save refactoring a lot of code later.

Thanks

**Comment by abdulimr** [comment]
I believe it is just a concern regarding increased latency.
This is helpful: https://github.com/vercel/next.js/discussions/16068#discussioncomment-49512

**Reply by kevanstannard** [reply]
Thanks @abdulimr 

Here's my understanding.

[1] If you are seeking best performance, then `getServerSideProps` should import the logic needed, and not call the API route.

[2] If you are seeking to decouple the page code from the API code, then calling the API routes is fine.

**Reply by abdulimr** [reply]
Yep [1] is the recommended approach.

I'm not too certain about [2]. Looking back, the documentation does mention that fetching from an external API is fine so I would be hesitant to do so for the internal `api` routes (apologies for not reading it entirely with my last comment). Although it is unlikely, there may be a technical concern as you've mentioned. I'm just wondering if your approach works as intended?

Hopefully someone can help with this!

**Comment by michrome** [comment]
If you directly import the logic then you know exactly what you’re calling against. This could be important if you’re using `getStaticProps` because otherwise you’ll call the version of API route that you pushed live in your _previous_ deploy. The commit you’re deploying now may have different logic that won’t yet be running at the API end point.

**Reply by kevanstannard** [reply]
Thanks @michrome, yes that makes sense for `getStaticProps`.

In this case I'm using `getServerSideProps` and I understand that the problem with calling a previous deployed API doesn't apply. Is that correct?
